{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class = \"alert alert-block alert-info\">\n",
    "    <h1><font color = \"red\">DISCLAIMER</font></h1>\n",
    "    <p>The following notebook it's highly based on the works <a href = \"https://www.kaggle.com/optimo/tabnetregressor-2-0-train-infer\">TabNetRegressor 2.0 [TRAIN + INFER]</a>, <a href = \"https://www.kaggle.com/liuhdme/moa-competition/data\">MOA competition</a> and <a href = \"https://www.kaggle.com/kushal1506/moa-pytorch-0-01859-rankgauss-pca-nn/data?select=train_targets_scored.csv\">\n",
    "MoA | Pytorch | 0.01859 | RankGauss | PCA | NN</a>, please check it out. I have to add that i don't make this notebook for \"upvotes\" but feedback.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = \"seagreen\">Preambule</font>\n",
    "\n",
    "I made this notebook to share some experiments (see the sections \"Experiments\") which could help to someone who don't want to wast their daily \"submissions\", but more importantly, to get feedback about what i could change to achive a better CV. Moreover, the easiness of TabNet to overfit the data it's disturbing. In the section \"Conclusion\" i share my opinion about the fine-tuning process of TabNet."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Installing Libraries</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "_cell_guid": "b1076dfc-b9ad-4769-8c92-a6c4dae69d19",
    "_kg_hide-output": true,
    "_uuid": "8f2839f25d086af736a60e9eeb907d3b93b6e0e5"
   },
   "outputs": [],
   "source": [
    "# # TabNet\n",
    "# !pip install --no-index --find-links /kaggle/input/pytorchtabnet/pytorch_tabnet-2.0.0-py3-none-any.whl pytorch-tabnet\n",
    "# # Iterative Stratification\n",
    "# !pip install /kaggle/input/iterative-stratification/iterative-stratification-master/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Loading Libraries</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "_cell_guid": "79c7e3d0-c299-4dcb-8224-4455121ee9b0",
    "_uuid": "d629ff2d2480ee46fbb7e2d37f6b5fab8052498a"
   },
   "outputs": [],
   "source": [
    "### General ###\n",
    "import os\n",
    "import sys\n",
    "import copy\n",
    "import tqdm\n",
    "import pickle\n",
    "import random\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "sys.path.append(\"../input/rank-gauss\")\n",
    "os.environ[\"CUDA_LAUNCH_BLOCKING\"] = '1'\n",
    "\n",
    "### Data Wrangling ###\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "from gauss_rank_scaler import GaussRankScaler\n",
    "\n",
    "### Data Visualization ###\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "plt.style.use(\"fivethirtyeight\")\n",
    "\n",
    "### Machine Learning ###\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.metrics import roc_auc_score, log_loss\n",
    "from sklearn.preprocessing import QuantileTransformer\n",
    "from sklearn.feature_selection import VarianceThreshold\n",
    "from iterstrat.ml_stratifiers import MultilabelStratifiedKFold\n",
    "\n",
    "### Deep Learning ###\n",
    "import torch\n",
    "from torch import nn\n",
    "import torch.optim as optim\n",
    "from torch.nn import functional as F\n",
    "from torch.nn.modules.loss import _WeightedLoss\n",
    "from torch.utils.data import DataLoader, Dataset\n",
    "from torch.optim.lr_scheduler import ReduceLROnPlateau\n",
    "# Tabnet \n",
    "from pytorch_tabnet.metrics import Metric\n",
    "from pytorch_tabnet.tab_model import TabNetRegressor\n",
    "\n",
    "### Make prettier the prints ###\n",
    "from colorama import Fore\n",
    "c_ = Fore.CYAN\n",
    "m_ = Fore.MAGENTA\n",
    "r_ = Fore.RED\n",
    "b_ = Fore.BLUE\n",
    "y_ = Fore.YELLOW\n",
    "g_ = Fore.GREEN"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Reproducibility</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "seed = 42\n",
    "\n",
    "def set_seed(seed):\n",
    "    random.seed(seed)\n",
    "    np.random.seed(seed)\n",
    "    torch.manual_seed(seed)\n",
    "    os.environ[\"PYTHONHASHSEED\"] = str(seed)\n",
    "    \n",
    "    if torch.cuda.is_available():\n",
    "        torch.cuda.manual_seed(seed)\n",
    "        torch.cuda.manual_seed_all(seed)\n",
    "        torch.backends.cudnn.deterministic = True\n",
    "        torch.backends.cudnn.benchmark = False\n",
    "set_seed(seed)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Configuration</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Parameters\n",
    "data_path = \"../input/lish-moa/\"\n",
    "no_ctl = True\n",
    "scale = \"rankgauss\"\n",
    "variance_threshould = 0.7\n",
    "decompo = \"PCA\"\n",
    "ncompo_genes = 80\n",
    "ncompo_cells = 10\n",
    "encoding = \"dummy\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Loading the Data</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "train = pd.read_csv(data_path + \"train_features.csv\")\n",
    "#train.drop(columns = [\"sig_id\"], inplace = True)\n",
    "\n",
    "targets = pd.read_csv(data_path + \"train_targets_scored.csv\")\n",
    "#train_targets_scored.drop(columns = [\"sig_id\"], inplace = True)\n",
    "\n",
    "#train_targets_nonscored = pd.read_csv(data_path + \"train_targets_nonscored.csv\")\n",
    "\n",
    "test = pd.read_csv(data_path + \"test_features.csv\")\n",
    "#test.drop(columns = [\"sig_id\"], inplace = True)\n",
    "\n",
    "submission = pd.read_csv(data_path + \"sample_submission.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = \"seagreen\">Preprocessing and Feature Engineering</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34m not_ctl\n"
     ]
    }
   ],
   "source": [
    "if no_ctl:\n",
    "    # cp_type == ctl_vehicle\n",
    "    print(b_, \"not_ctl\")\n",
    "    train = train[train[\"cp_type\"] != \"ctl_vehicle\"]\n",
    "    test = test[test[\"cp_type\"] != \"ctl_vehicle\"]\n",
    "    targets = targets.iloc[train.index]\n",
    "    train.reset_index(drop = True, inplace = True)\n",
    "    test.reset_index(drop = True, inplace = True)\n",
    "    targets.reset_index(drop = True, inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Distributions Before Rank Gauss and PCA</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def distributions(num, graphs, items, features, gorc):\n",
    "    \"\"\"\n",
    "    Plot the distributions of gene expression or cell viability data\n",
    "    \"\"\"\n",
    "    for i in range(0, num - 1, 7):\n",
    "        if i >= 3:\n",
    "            break\n",
    "        idxs = list(np.array([0, 1, 2, 3, 4, 5, 6]) + i)\n",
    "    \n",
    "        fig, axs = plt.subplots(1, 7, sharey = True)\n",
    "        for k, item in enumerate(idxs):\n",
    "            if item >= items:\n",
    "                break\n",
    "            graph = sns.distplot(train[features].values[:, item], ax = axs[k])\n",
    "            graph.set_title(f\"{gorc}-{item}\")\n",
    "            graphs.append(graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "GENES = [col for col in train.columns if col.startswith(\"g-\")]\n",
    "CELLS = [col for col in train.columns if col.startswith(\"c-\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### <font color = \"green\">Distributions of the Train Set</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Rank Gauss Process</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle\n",
    "\n",
    "data_all = pd.concat([train, test], ignore_index = True)\n",
    "cols_numeric = [feat for feat in list(data_all.columns) if feat not in [\"sig_id\", \"cp_type\", \"cp_time\", \"cp_dose\"]]\n",
    "mask = (data_all[cols_numeric].var() >= variance_threshould).values\n",
    "\n",
    "file = open('mask.pickle', 'wb')\n",
    "pickle.dump(mask, file)\n",
    "file.close()\n",
    "\n",
    "\n",
    "\n",
    "tmp = data_all[cols_numeric].loc[:, mask]\n",
    "data_all = pd.concat([data_all[[\"sig_id\", \"cp_type\", \"cp_time\", \"cp_dose\"]], tmp], axis = 1)\n",
    "cols_numeric = [feat for feat in list(data_all.columns) if feat not in [\"sig_id\", \"cp_type\", \"cp_time\", \"cp_dose\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34m Rank Gauss\n"
     ]
    }
   ],
   "source": [
    "def scale_minmax(col):\n",
    "    return (col - col.min()) / (col.max() - col.min())\n",
    "\n",
    "def scale_norm(col):\n",
    "    return (col - col.mean()) / col.std()\n",
    "\n",
    "if scale == \"boxcox\":\n",
    "    print(b_, \"boxcox\")\n",
    "    data_all[cols_numeric] = data_all[cols_numeric].apply(scale_minmax, axis = 0)\n",
    "    trans = []\n",
    "    for feat in cols_numeric:\n",
    "        trans_var, lambda_var = stats.boxcox(data_all[feat].dropna() + 1)\n",
    "        trans.append(scale_minmax(trans_var))\n",
    "    data_all[cols_numeric] = np.asarray(trans).T\n",
    "    \n",
    "elif scale == \"norm\":\n",
    "    print(b_, \"norm\")\n",
    "    data_all[cols_numeric] = data_all[cols_numeric].apply(scale_norm, axis = 0)\n",
    "    \n",
    "elif scale == \"minmax\":\n",
    "    print(b_, \"minmax\")\n",
    "    data_all[cols_numeric] = data_all[cols_numeric].apply(scale_minmax, axis = 0)\n",
    "    \n",
    "elif scale == \"rankgauss\":\n",
    "    ### Rank Gauss ###\n",
    "    print(b_, \"Rank Gauss\")\n",
    "    guassscaler = GaussRankScaler().fit(data_all[cols_numeric])\n",
    "    \n",
    "    file = open('guassscaler.pickle', 'wb')\n",
    "    pickle.dump(guassscaler, file)\n",
    "    file.close()\n",
    "\n",
    "    \n",
    "    data_all[cols_numeric] = guassscaler.transform(data_all[cols_numeric])\n",
    "\n",
    "#     data_all[cols_numeric] = scaler.fit_transform(data_all[cols_numeric])\n",
    "    \n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Principal Component Analysis</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34m PCA\n"
     ]
    }
   ],
   "source": [
    "# PCA\n",
    "if decompo == \"PCA\":\n",
    "    print(b_, \"PCA\")\n",
    "    GENES = [col for col in data_all.columns if col.startswith(\"g-\")]\n",
    "    CELLS = [col for col in data_all.columns if col.startswith(\"c-\")]\n",
    "    \n",
    "#     pca_genes = PCA(n_components = ncompo_genes,\n",
    "#                     random_state = seed).fit_transform(data_all[GENES])\n",
    "    gpca = PCA(n_components = ncompo_genes,\n",
    "                    random_state = seed).fit(data_all[GENES])\n",
    "    pca_genes = gpca.transform(data_all[GENES])\n",
    "                \n",
    "    \n",
    "    file = open('gpca.pickle', 'wb')\n",
    "    pickle.dump(gpca, file)\n",
    "    file.close()\n",
    "\n",
    "\n",
    "#     pca_cells = PCA(n_components = ncompo_cells,\n",
    "#                     random_state = seed).fit_transform(data_all[CELLS])\n",
    "    cpca = PCA(n_components = ncompo_cells,\n",
    "                    random_state = seed).fit(data_all[CELLS])\n",
    "    pca_cells = cpca.transform(data_all[CELLS])\n",
    "        \n",
    "    \n",
    "    file = open('cpca.pickle', 'wb')\n",
    "    pickle.dump(cpca, file)\n",
    "    file.close()\n",
    "\n",
    "\n",
    "    \n",
    "    pca_genes = pd.DataFrame(pca_genes, columns = [f\"pca_g-{i}\" for i in range(ncompo_genes)])\n",
    "    pca_cells = pd.DataFrame(pca_cells, columns = [f\"pca_c-{i}\" for i in range(ncompo_cells)])\n",
    "    data_all = pd.concat([data_all, pca_genes, pca_cells], axis = 1)\n",
    "else:\n",
    "    pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">One Hot</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34m One-Hot\n"
     ]
    }
   ],
   "source": [
    "# Encoding\n",
    "if encoding == \"lb\":\n",
    "    print(b_, \"Label Encoding\")\n",
    "    for feat in [\"cp_time\", \"cp_dose\"]:\n",
    "        data_all[feat] = LabelEncoder().fit_transform(data_all[feat])\n",
    "elif encoding == \"dummy\":\n",
    "    print(b_, \"One-Hot\")\n",
    "    data_all = pd.get_dummies(data_all, columns = [\"cp_time\", \"cp_dose\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 5/5 [00:05<00:00,  1.13s/it]\n"
     ]
    }
   ],
   "source": [
    "GENES = [col for col in data_all.columns if col.startswith(\"g-\")]\n",
    "CELLS = [col for col in data_all.columns if col.startswith(\"c-\")]\n",
    "\n",
    "for stats in tqdm.tqdm([\"sum\", \"mean\", \"std\", \"kurt\", \"skew\"]):\n",
    "    data_all[\"g_\" + stats] = getattr(data_all[GENES], stats)(axis = 1)\n",
    "    data_all[\"c_\" + stats] = getattr(data_all[CELLS], stats)(axis = 1)    \n",
    "    data_all[\"gc_\" + stats] = getattr(data_all[GENES + CELLS], stats)(axis = 1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Distributions After Rank Gauss and PCA</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def distributions(num, graphs, items, features, gorc):\n",
    "    \"\"\"\n",
    "    Plot the distributions of gene expression or cell viability data\n",
    "    \"\"\"\n",
    "    for i in range(0, num - 1, 7):\n",
    "        if i >= 3:\n",
    "            break\n",
    "        idxs = list(np.array([0, 1, 2, 3, 4, 5, 6]) + i)\n",
    "    \n",
    "        fig, axs = plt.subplots(1, 7, sharey = True)\n",
    "        for k, item in enumerate(idxs):\n",
    "            if item >= items:\n",
    "                break\n",
    "            graph = sns.distplot(data_all[features].values[:, item], ax = axs[k])\n",
    "            graph.set_title(f\"{gorc}-{item}\")\n",
    "            graphs.append(graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0        id_000644bb2\n",
       "1        id_000779bfc\n",
       "2        id_000a6266a\n",
       "3        id_0015fd391\n",
       "4        id_001626bd3\n",
       "             ...     \n",
       "21943    id_fff8c2444\n",
       "21944    id_fffb1ceed\n",
       "21945    id_fffb70c0c\n",
       "21946    id_fffcb9e7c\n",
       "21947    id_ffffdd77b\n",
       "Name: sig_id, Length: 21948, dtype: object"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df_ID = data_all.sig_id[: train.shape[0]]\n",
    "train_df_ID"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# train_df and test_df\n",
    "train_df_ID = data_all.sig_id[: train.shape[0]]\n",
    "features_to_drop = [\"sig_id\", \"cp_type\"]\n",
    "data_all.drop(features_to_drop, axis = 1, inplace = True)\n",
    "try:\n",
    "    targets.drop(\"sig_id\", axis = 1, inplace = True)\n",
    "except:\n",
    "    pass\n",
    "train_df = data_all[: train.shape[0]]\n",
    "train_df.reset_index(drop = True, inplace = True)\n",
    "# The following line it's a bad practice in my opinion, targets on train set\n",
    "#train_df = pd.concat([train_df, targets], axis = 1)\n",
    "test_df = data_all[train_df.shape[0]: ]\n",
    "test_df.reset_index(drop = True, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mtrain_df.shape: \u001b[31m(21948, 947)\n",
      "\u001b[34mtest_df.shape: \u001b[31m(3624, 947)\n"
     ]
    }
   ],
   "source": [
    "print(f\"{b_}train_df.shape: {r_}{train_df.shape}\")\n",
    "print(f\"{b_}test_df.shape: {r_}{test_df.shape}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34mX_test.shape: \u001b[31m(3624, 947)\n"
     ]
    }
   ],
   "source": [
    "X_test = test_df.values\n",
    "print(f\"{b_}X_test.shape: {r_}{X_test.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = \"seagreen\">Experiments</font>\n",
    "\n",
    "I just want to point that the [original work](https://www.kaggle.com/optimo/tabnetregressor-2-0-train-infer) achive a CV of 0.015532370835690834 and a LB score of 0.01864. Some of the experiments that i made with their changes:\n",
    "\n",
    "\n",
    "- CV: 0.01543560538566987, LB: 0.01858, best LB that i could achive, changes\n",
    "    - `n_a` = 32 instead of 24\n",
    "    - `n_d` = 32 instead of 24\n",
    "- CV: 0.015282077428722094, LB: 0.01862, best CV that i could achive, changes (Version 5):\n",
    "    - `n_a` = 32 instead of 24\n",
    "    - `n_d` = 32 instead of 24\n",
    "    - `virtual_batch_size` = 32, instead of 128\n",
    "    - `seed` = 42 instead of 0\n",
    "- CV: 0.015330138325308062, LB: 01864, the same LB that the original but better CV, changes:\n",
    "    - `n_a` = 32 instead of 24\n",
    "    - `n_d` = 32 instead of 24\n",
    "    - `virtual_batch_size` = 64, instead of 128\n",
    "    - `batch_size` = 512, instead of 1024\n",
    "- CV: 0.015361751699863063, LB: 0.01863, better LB and CV than the original, changes:\n",
    "    - `n_a` = 32 instead of 24\n",
    "    - `n_d` = 32 instead of 24\n",
    "    - `virtual_batch_size` = 64, instead of 128\n",
    "- CV: 0.015529925324634975, LB: 0.01865, changes:\n",
    "    - `n_a` = 48 instead of 24\n",
    "    - `n_d` = 48 instead of 24\n",
    "- CV: 0.015528553520924939, LB: 0.01868, changes:\n",
    "    - `n_a` = 12 instead of 24\n",
    "    - `n_d` = 12 instead of 24\n",
    "- CV: 0.015870202970324317, LB: 0.01876, worst CV and LB score, changes:\n",
    "    - `n_a` = 12 instead of 24\n",
    "    - `n_d` = 12 instead of 24\n",
    "    - `batch_size` = 2048, instead of 1024\n",
    "    \n",
    "    \n",
    "As you can see if `batch_size` < 1024 and > 1024 give worst results. Something similar happens with `n_a` and `n_d`, if their values are lower or higher than 32 the results are worst.\n",
    "\n",
    "\n",
    "## <font color = \"green\">Versions</font>\n",
    "\n",
    "- **Version 5**: I added the `seed` parameter to the TabNet model.\n",
    "- **Version 6**: I changed the `virtual_batch_size` to 24\n",
    "    - CV: 0.01532900616425282, LB: 0.01862, changes:\n",
    "        - `n_a` = 32 instead of 24\n",
    "        - `n_d` = 32 instead of 24\n",
    "        - `virtual_batch_size` = 24, instead of 128\n",
    "        - `seed` = 42 instead of 0\n",
    "- **Version 7**: PCA, Rank Gauss"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = \"seagreen\">Modeling</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Model Parameters</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAX_EPOCH = 200\n",
    "# n_d and n_a are different from the original work, 32 instead of 24\n",
    "# This is the first change in the code from the original\n",
    "tabnet_params = dict(\n",
    "    n_d = 32,\n",
    "    n_a = 32,\n",
    "    n_steps = 1,\n",
    "    gamma = 1.3,\n",
    "    lambda_sparse = 0,\n",
    "    optimizer_fn = optim.Adam,\n",
    "    optimizer_params = dict(lr = 2e-2, weight_decay = 1e-5),\n",
    "    mask_type = \"entmax\",\n",
    "    scheduler_params = dict(\n",
    "        mode = \"min\", patience = 5, min_lr = 1e-5, factor = 0.9),\n",
    "    scheduler_fn = ReduceLROnPlateau,\n",
    "    seed = seed,\n",
    "    verbose = 10\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## <font color = \"green\">Custom Metric</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "class LogitsLogLoss(Metric):\n",
    "    \"\"\"\n",
    "    LogLoss with sigmoid applied\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self):\n",
    "        self._name = \"logits_ll\"\n",
    "        self._maximize = False\n",
    "\n",
    "    def __call__(self, y_true, y_pred):\n",
    "        \"\"\"\n",
    "        Compute LogLoss of predictions.\n",
    "\n",
    "        Parameters\n",
    "        ----------\n",
    "        y_true: np.ndarray\n",
    "            Target matrix or vector\n",
    "        y_score: np.ndarray\n",
    "            Score matrix or vector\n",
    "\n",
    "        Returns\n",
    "        -------\n",
    "            float\n",
    "            LogLoss of predictions vs targets.\n",
    "        \"\"\"\n",
    "        logits = 1 / (1 + np.exp(-y_pred))\n",
    "        aux = (1 - y_true) * np.log(1 - logits + 1e-15) + y_true * np.log(logits + 1e-15)\n",
    "        return np.mean(-aux)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g-0</th>\n",
       "      <th>g-2</th>\n",
       "      <th>g-3</th>\n",
       "      <th>g-4</th>\n",
       "      <th>g-5</th>\n",
       "      <th>g-6</th>\n",
       "      <th>g-7</th>\n",
       "      <th>g-8</th>\n",
       "      <th>g-9</th>\n",
       "      <th>g-10</th>\n",
       "      <th>...</th>\n",
       "      <th>gc_mean</th>\n",
       "      <th>g_std</th>\n",
       "      <th>c_std</th>\n",
       "      <th>gc_std</th>\n",
       "      <th>g_kurt</th>\n",
       "      <th>c_kurt</th>\n",
       "      <th>gc_kurt</th>\n",
       "      <th>g_skew</th>\n",
       "      <th>c_skew</th>\n",
       "      <th>gc_skew</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.731433</td>\n",
       "      <td>-0.254104</td>\n",
       "      <td>-0.615315</td>\n",
       "      <td>-0.194236</td>\n",
       "      <td>-0.772257</td>\n",
       "      <td>-1.061067</td>\n",
       "      <td>0.005623</td>\n",
       "      <td>0.513989</td>\n",
       "      <td>-0.130963</td>\n",
       "      <td>1.094414</td>\n",
       "      <td>...</td>\n",
       "      <td>0.045182</td>\n",
       "      <td>0.600465</td>\n",
       "      <td>0.506976</td>\n",
       "      <td>0.602876</td>\n",
       "      <td>-0.006781</td>\n",
       "      <td>-0.250230</td>\n",
       "      <td>-0.025857</td>\n",
       "      <td>-0.010803</td>\n",
       "      <td>0.118667</td>\n",
       "      <td>-0.039037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.017020</td>\n",
       "      <td>0.213820</td>\n",
       "      <td>0.027107</td>\n",
       "      <td>0.849834</td>\n",
       "      <td>0.482972</td>\n",
       "      <td>0.217779</td>\n",
       "      <td>0.363904</td>\n",
       "      <td>-0.298420</td>\n",
       "      <td>0.730029</td>\n",
       "      <td>-0.808914</td>\n",
       "      <td>...</td>\n",
       "      <td>0.045437</td>\n",
       "      <td>0.588216</td>\n",
       "      <td>0.421628</td>\n",
       "      <td>0.584621</td>\n",
       "      <td>-0.048549</td>\n",
       "      <td>0.111352</td>\n",
       "      <td>-0.096328</td>\n",
       "      <td>0.096908</td>\n",
       "      <td>-0.043717</td>\n",
       "      <td>0.003906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.462792</td>\n",
       "      <td>0.981712</td>\n",
       "      <td>-0.091777</td>\n",
       "      <td>-0.039121</td>\n",
       "      <td>1.055526</td>\n",
       "      <td>0.161005</td>\n",
       "      <td>0.242395</td>\n",
       "      <td>0.050070</td>\n",
       "      <td>0.975415</td>\n",
       "      <td>-0.385579</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.022895</td>\n",
       "      <td>0.643043</td>\n",
       "      <td>0.458371</td>\n",
       "      <td>0.623742</td>\n",
       "      <td>-0.366360</td>\n",
       "      <td>-0.051865</td>\n",
       "      <td>-0.285094</td>\n",
       "      <td>-0.015580</td>\n",
       "      <td>0.379548</td>\n",
       "      <td>0.011925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.531916</td>\n",
       "      <td>-0.269127</td>\n",
       "      <td>0.455184</td>\n",
       "      <td>1.595325</td>\n",
       "      <td>-0.621171</td>\n",
       "      <td>-1.509230</td>\n",
       "      <td>0.206249</td>\n",
       "      <td>-0.066407</td>\n",
       "      <td>-0.889590</td>\n",
       "      <td>-0.656089</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.096822</td>\n",
       "      <td>0.722179</td>\n",
       "      <td>0.392981</td>\n",
       "      <td>0.724922</td>\n",
       "      <td>-0.877465</td>\n",
       "      <td>4.444686</td>\n",
       "      <td>-0.907331</td>\n",
       "      <td>0.108122</td>\n",
       "      <td>2.010409</td>\n",
       "      <td>0.272977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.341668</td>\n",
       "      <td>0.808567</td>\n",
       "      <td>0.604615</td>\n",
       "      <td>0.951946</td>\n",
       "      <td>-0.630950</td>\n",
       "      <td>-0.243933</td>\n",
       "      <td>-0.103294</td>\n",
       "      <td>-0.684346</td>\n",
       "      <td>0.741863</td>\n",
       "      <td>-0.175749</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024536</td>\n",
       "      <td>0.765814</td>\n",
       "      <td>0.471472</td>\n",
       "      <td>0.743453</td>\n",
       "      <td>-0.353646</td>\n",
       "      <td>-0.477550</td>\n",
       "      <td>-0.243094</td>\n",
       "      <td>-0.127465</td>\n",
       "      <td>0.191181</td>\n",
       "      <td>-0.198329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 947 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        g-0       g-2       g-3       g-4       g-5       g-6       g-7  \\\n",
       "0  0.731433 -0.254104 -0.615315 -0.194236 -0.772257 -1.061067  0.005623   \n",
       "1  0.017020  0.213820  0.027107  0.849834  0.482972  0.217779  0.363904   \n",
       "2  0.462792  0.981712 -0.091777 -0.039121  1.055526  0.161005  0.242395   \n",
       "3 -0.531916 -0.269127  0.455184  1.595325 -0.621171 -1.509230  0.206249   \n",
       "4 -0.341668  0.808567  0.604615  0.951946 -0.630950 -0.243933 -0.103294   \n",
       "\n",
       "        g-8       g-9      g-10  ...   gc_mean     g_std     c_std    gc_std  \\\n",
       "0  0.513989 -0.130963  1.094414  ...  0.045182  0.600465  0.506976  0.602876   \n",
       "1 -0.298420  0.730029 -0.808914  ...  0.045437  0.588216  0.421628  0.584621   \n",
       "2  0.050070  0.975415 -0.385579  ... -0.022895  0.643043  0.458371  0.623742   \n",
       "3 -0.066407 -0.889590 -0.656089  ... -0.096822  0.722179  0.392981  0.724922   \n",
       "4 -0.684346  0.741863 -0.175749  ...  0.024536  0.765814  0.471472  0.743453   \n",
       "\n",
       "     g_kurt    c_kurt   gc_kurt    g_skew    c_skew   gc_skew  \n",
       "0 -0.006781 -0.250230 -0.025857 -0.010803  0.118667 -0.039037  \n",
       "1 -0.048549  0.111352 -0.096328  0.096908 -0.043717  0.003906  \n",
       "2 -0.366360 -0.051865 -0.285094 -0.015580  0.379548  0.011925  \n",
       "3 -0.877465  4.444686 -0.907331  0.108122  2.010409  0.272977  \n",
       "4 -0.353646 -0.477550 -0.243094 -0.127465  0.191181 -0.198329  \n",
       "\n",
       "[5 rows x 947 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = \"seagreen\">Training</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34m FOLDS:  \u001b[31m 1\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_002452c7e' 'id_003603254' 'id_003fdd734' ... 'id_ffa08c24c'\n",
      " 'id_ffab8a71d' 'id_fff183968']\n",
      "Device used : cuda\n",
      "epoch 0  | loss: 0.29866 | val_logits_ll: 0.02837 |  0:00:02s\n",
      "epoch 10 | loss: 0.01854 | val_logits_ll: 0.02076 |  0:00:25s\n",
      "epoch 20 | loss: 0.01731 | val_logits_ll: 0.01741 |  0:00:48s\n",
      "epoch 30 | loss: 0.01687 | val_logits_ll: 0.01813 |  0:01:11s\n",
      "epoch 40 | loss: 0.01674 | val_logits_ll: 0.01709 |  0:01:41s\n",
      "epoch 50 | loss: 0.0164  | val_logits_ll: 0.01676 |  0:02:11s\n",
      "epoch 60 | loss: 0.01618 | val_logits_ll: 0.01674 |  0:02:41s\n",
      "epoch 70 | loss: 0.0157  | val_logits_ll: 0.01683 |  0:03:05s\n",
      "epoch 80 | loss: 0.01558 | val_logits_ll: 0.01672 |  0:03:35s\n",
      "epoch 90 | loss: 0.01517 | val_logits_ll: 0.01666 |  0:03:57s\n",
      "\n",
      "Early stopping occured at epoch 97 with best_epoch = 77 and best_val_logits_ll = 0.01653\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n",
      "\u001b[34m FOLDS:  \u001b[31m 2\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_0023f063e' 'id_002e08ff8' 'id_0036b0342' ... 'id_fff11dbf5'\n",
      " 'id_fff485da4' 'id_fffb1ceed']\n",
      "Device used : cuda\n",
      "epoch 0  | loss: 0.29878 | val_logits_ll: 0.02878 |  0:00:03s\n",
      "epoch 10 | loss: 0.01869 | val_logits_ll: 0.01883 |  0:00:33s\n",
      "epoch 20 | loss: 0.01738 | val_logits_ll: 0.01969 |  0:01:03s\n",
      "epoch 30 | loss: 0.017   | val_logits_ll: 0.0171  |  0:01:29s\n",
      "epoch 40 | loss: 0.01672 | val_logits_ll: 0.01752 |  0:01:59s\n",
      "epoch 50 | loss: 0.01649 | val_logits_ll: 0.01691 |  0:02:26s\n",
      "epoch 60 | loss: 0.01645 | val_logits_ll: 0.0169  |  0:02:54s\n",
      "epoch 70 | loss: 0.01596 | val_logits_ll: 0.01672 |  0:03:22s\n",
      "epoch 80 | loss: 0.01569 | val_logits_ll: 0.01664 |  0:03:51s\n",
      "epoch 90 | loss: 0.01561 | val_logits_ll: 0.01672 |  0:04:19s\n",
      "\n",
      "Early stopping occured at epoch 92 with best_epoch = 72 and best_val_logits_ll = 0.01647\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n",
      "\u001b[34m FOLDS:  \u001b[31m 3\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_0024bcd70' 'id_003b43239' 'id_005c3cb48' ... 'id_ff8e3f7d3'\n",
      " 'id_ff9328b5e' 'id_ffc9721f4']\n",
      "Device used : cuda\n",
      "epoch 0  | loss: 0.29794 | val_logits_ll: 0.02857 |  0:00:03s\n",
      "epoch 10 | loss: 0.01897 | val_logits_ll: 0.01876 |  0:00:30s\n",
      "epoch 20 | loss: 0.01751 | val_logits_ll: 0.0233  |  0:00:58s\n",
      "epoch 30 | loss: 0.01731 | val_logits_ll: 0.01755 |  0:01:28s\n",
      "epoch 40 | loss: 0.01673 | val_logits_ll: 0.0177  |  0:01:58s\n",
      "epoch 50 | loss: 0.01666 | val_logits_ll: 0.01709 |  0:02:28s\n",
      "epoch 60 | loss: 0.01645 | val_logits_ll: 0.01694 |  0:02:56s\n",
      "epoch 70 | loss: 0.01648 | val_logits_ll: 0.01695 |  0:03:23s\n",
      "epoch 80 | loss: 0.01611 | val_logits_ll: 0.01693 |  0:03:51s\n",
      "epoch 90 | loss: 0.01584 | val_logits_ll: 0.01687 |  0:04:20s\n",
      "epoch 100| loss: 0.01556 | val_logits_ll: 0.0168  |  0:04:50s\n",
      "epoch 110| loss: 0.01553 | val_logits_ll: 0.01669 |  0:05:19s\n",
      "epoch 120| loss: 0.01505 | val_logits_ll: 0.01672 |  0:05:47s\n",
      "\n",
      "Early stopping occured at epoch 125 with best_epoch = 105 and best_val_logits_ll = 0.01657\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n",
      "\u001b[34m FOLDS:  \u001b[31m 4\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_000a6266a' 'id_001626bd3' 'id_002fb9c19' ... 'id_ffe18c87f'\n",
      " 'id_fff6df1c5' 'id_fff8c2444']\n",
      "Device used : cuda\n",
      "epoch 0  | loss: 0.29909 | val_logits_ll: 0.02854 |  0:00:03s\n",
      "epoch 10 | loss: 0.01887 | val_logits_ll: 0.01864 |  0:00:31s\n",
      "epoch 20 | loss: 0.01739 | val_logits_ll: 0.01961 |  0:00:59s\n",
      "epoch 30 | loss: 0.01689 | val_logits_ll: 0.0175  |  0:01:49s\n",
      "epoch 40 | loss: 0.01647 | val_logits_ll: 0.01678 |  0:02:55s\n",
      "epoch 50 | loss: 0.01615 | val_logits_ll: 0.01674 |  0:04:03s\n",
      "epoch 60 | loss: 0.01598 | val_logits_ll: 0.0166  |  0:05:13s\n",
      "epoch 70 | loss: 0.01595 | val_logits_ll: 0.01682 |  0:06:20s\n",
      "epoch 80 | loss: 0.0156  | val_logits_ll: 0.01648 |  0:07:18s\n",
      "epoch 90 | loss: 0.01532 | val_logits_ll: 0.01657 |  0:08:27s\n",
      "epoch 100| loss: 0.01504 | val_logits_ll: 0.01665 |  0:09:37s\n",
      "\n",
      "Early stopping occured at epoch 106 with best_epoch = 86 and best_val_logits_ll = 0.01646\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n",
      "\u001b[34m FOLDS:  \u001b[31m 5\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_001762a82' 'id_0025c5949' 'id_002d31e2c' ... 'id_ffe6452ba'\n",
      " 'id_fff3976bd' 'id_fff7e6992']\n",
      "Device used : cuda\n",
      "epoch 0  | loss: 0.2989  | val_logits_ll: 0.02905 |  0:00:02s\n",
      "epoch 10 | loss: 0.01871 | val_logits_ll: 0.01867 |  0:01:10s\n",
      "epoch 20 | loss: 0.01742 | val_logits_ll: 0.02055 |  0:02:14s\n",
      "epoch 30 | loss: 0.01703 | val_logits_ll: 0.01738 |  0:03:25s\n",
      "epoch 40 | loss: 0.0167  | val_logits_ll: 0.0172  |  0:04:20s\n",
      "epoch 50 | loss: 0.01661 | val_logits_ll: 0.01715 |  0:04:57s\n",
      "epoch 60 | loss: 0.01642 | val_logits_ll: 0.01716 |  0:05:25s\n",
      "epoch 70 | loss: 0.01628 | val_logits_ll: 0.01688 |  0:05:53s\n",
      "epoch 80 | loss: 0.01613 | val_logits_ll: 0.01687 |  0:06:21s\n",
      "epoch 90 | loss: 0.01585 | val_logits_ll: 0.01672 |  0:06:50s\n",
      "epoch 100| loss: 0.01563 | val_logits_ll: 0.01695 |  0:07:19s\n",
      "epoch 110| loss: 0.0154  | val_logits_ll: 0.01674 |  0:07:48s\n",
      "epoch 120| loss: 0.01519 | val_logits_ll: 0.0167  |  0:08:16s\n",
      "epoch 130| loss: 0.01486 | val_logits_ll: 0.01678 |  0:08:45s\n",
      "\n",
      "Early stopping occured at epoch 135 with best_epoch = 115 and best_val_logits_ll = 0.01655\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n",
      "\u001b[34m FOLDS:  \u001b[31m 6\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_000779bfc' 'id_0015fd391' 'id_008c40337' ... 'id_fff1a3307'\n",
      " 'id_fffb70c0c' 'id_fffcb9e7c']\n",
      "Device used : cuda\n",
      "epoch 0  | loss: 0.29884 | val_logits_ll: 0.02966 |  0:00:03s\n",
      "epoch 10 | loss: 0.01894 | val_logits_ll: 0.01865 |  0:00:31s\n",
      "epoch 20 | loss: 0.01744 | val_logits_ll: 0.01951 |  0:01:00s\n",
      "epoch 30 | loss: 0.01706 | val_logits_ll: 0.01696 |  0:01:28s\n",
      "epoch 40 | loss: 0.01684 | val_logits_ll: 0.01872 |  0:01:56s\n",
      "epoch 50 | loss: 0.01669 | val_logits_ll: 0.01699 |  0:02:27s\n",
      "epoch 60 | loss: 0.01622 | val_logits_ll: 0.0166  |  0:02:55s\n",
      "epoch 70 | loss: 0.01619 | val_logits_ll: 0.0166  |  0:03:22s\n",
      "epoch 80 | loss: 0.01593 | val_logits_ll: 0.01655 |  0:03:52s\n",
      "epoch 90 | loss: 0.0156  | val_logits_ll: 0.01647 |  0:04:21s\n",
      "epoch 100| loss: 0.01542 | val_logits_ll: 0.01651 |  0:04:51s\n",
      "epoch 110| loss: 0.01513 | val_logits_ll: 0.01637 |  0:05:20s\n",
      "epoch 120| loss: 0.01487 | val_logits_ll: 0.01661 |  0:05:51s\n",
      "\n",
      "Early stopping occured at epoch 129 with best_epoch = 109 and best_val_logits_ll = 0.01635\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n",
      "\u001b[34m FOLDS:  \u001b[31m 7\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_0020d0484' 'id_005c3ee1a' 'id_009379beb' ... 'id_ff808ce8b'\n",
      " 'id_ffaac2669' 'id_ffffdd77b']\n",
      "Device used : cuda\n",
      "epoch 0  | loss: 0.29627 | val_logits_ll: 0.02848 |  0:00:03s\n",
      "epoch 10 | loss: 0.01924 | val_logits_ll: 0.01902 |  0:00:32s\n",
      "epoch 20 | loss: 0.01751 | val_logits_ll: 0.01764 |  0:01:02s\n",
      "epoch 30 | loss: 0.01701 | val_logits_ll: 0.01723 |  0:01:31s\n",
      "epoch 40 | loss: 0.01672 | val_logits_ll: 0.01688 |  0:02:02s\n",
      "epoch 50 | loss: 0.01644 | val_logits_ll: 0.01676 |  0:02:31s\n",
      "epoch 60 | loss: 0.0163  | val_logits_ll: 0.01664 |  0:03:01s\n",
      "epoch 70 | loss: 0.01604 | val_logits_ll: 0.01663 |  0:03:31s\n",
      "epoch 80 | loss: 0.01585 | val_logits_ll: 0.01655 |  0:04:30s\n",
      "epoch 90 | loss: 0.01578 | val_logits_ll: 0.0165  |  0:05:40s\n",
      "epoch 100| loss: 0.01559 | val_logits_ll: 0.01645 |  0:06:53s\n",
      "epoch 110| loss: 0.01519 | val_logits_ll: 0.01648 |  0:07:53s\n",
      "epoch 120| loss: 0.01506 | val_logits_ll: 0.01648 |  0:08:49s\n",
      "\n",
      "Early stopping occured at epoch 122 with best_epoch = 102 and best_val_logits_ll = 0.01633\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n",
      "\u001b[34m FOLDS:  \u001b[31m 8\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_000644bb2' 'id_001bd861f' 'id_003d4b6f0' ... 'id_ffd1fd119'\n",
      " 'id_fff506720' 'id_fff790a17']\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Device used : cuda\n",
      "epoch 0  | loss: 0.29638 | val_logits_ll: 0.02877 |  0:00:08s\n",
      "epoch 10 | loss: 0.01875 | val_logits_ll: 0.01953 |  0:01:21s\n",
      "epoch 20 | loss: 0.01739 | val_logits_ll: 0.01992 |  0:02:25s\n",
      "epoch 30 | loss: 0.01687 | val_logits_ll: 0.01753 |  0:03:35s\n",
      "epoch 40 | loss: 0.01663 | val_logits_ll: 0.01704 |  0:04:45s\n",
      "epoch 50 | loss: 0.01632 | val_logits_ll: 0.017   |  0:05:54s\n",
      "epoch 60 | loss: 0.01615 | val_logits_ll: 0.01694 |  0:06:57s\n",
      "epoch 70 | loss: 0.01588 | val_logits_ll: 0.01669 |  0:07:52s\n",
      "\n",
      "Early stopping occured at epoch 77 with best_epoch = 57 and best_val_logits_ll = 0.01659\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n",
      "\u001b[34m FOLDS:  \u001b[31m 9\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_00505b3c8' 'id_0062bfc63' 'id_0092c905e' ... 'id_ffd241f1c'\n",
      " 'id_ffdd24c81' 'id_fff26b3c2']\n",
      "Device used : cuda\n",
      "epoch 0  | loss: 0.29643 | val_logits_ll: 0.02784 |  0:00:02s\n",
      "epoch 10 | loss: 0.01908 | val_logits_ll: 0.01957 |  0:00:32s\n",
      "epoch 20 | loss: 0.01752 | val_logits_ll: 0.02068 |  0:01:00s\n",
      "epoch 30 | loss: 0.01706 | val_logits_ll: 0.01784 |  0:01:31s\n",
      "epoch 40 | loss: 0.01676 | val_logits_ll: 0.01725 |  0:02:00s\n",
      "epoch 50 | loss: 0.01642 | val_logits_ll: 0.01691 |  0:02:30s\n",
      "epoch 60 | loss: 0.01631 | val_logits_ll: 0.01679 |  0:02:59s\n",
      "epoch 70 | loss: 0.0161  | val_logits_ll: 0.0167  |  0:03:29s\n",
      "epoch 80 | loss: 0.0159  | val_logits_ll: 0.01663 |  0:03:59s\n",
      "epoch 90 | loss: 0.01566 | val_logits_ll: 0.01674 |  0:04:28s\n",
      "epoch 100| loss: 0.01542 | val_logits_ll: 0.01663 |  0:04:58s\n",
      "epoch 110| loss: 0.0151  | val_logits_ll: 0.01669 |  0:05:27s\n",
      "epoch 120| loss: 0.01492 | val_logits_ll: 0.01679 |  0:05:58s\n",
      "epoch 130| loss: 0.01456 | val_logits_ll: 0.01673 |  0:06:27s\n",
      "\n",
      "Early stopping occured at epoch 137 with best_epoch = 117 and best_val_logits_ll = 0.01658\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n",
      "\u001b[34m FOLDS:  \u001b[31m 10\n",
      "\u001b[32m ************************************************************ \u001b[36m\n",
      "['id_00224bf20' 'id_00289dc6e' 'id_0039a2ff9' ... 'id_ff68e852c'\n",
      " 'id_ff7b158fa' 'id_ffcb7e365']\n",
      "Device used : cuda\n",
      "epoch 0  | loss: 0.29968 | val_logits_ll: 0.02819 |  0:00:03s\n",
      "epoch 10 | loss: 0.01909 | val_logits_ll: 0.01896 |  0:00:32s\n",
      "epoch 20 | loss: 0.01741 | val_logits_ll: 0.01819 |  0:01:01s\n",
      "epoch 30 | loss: 0.01699 | val_logits_ll: 0.01849 |  0:01:32s\n",
      "epoch 40 | loss: 0.01683 | val_logits_ll: 0.01708 |  0:02:01s\n",
      "epoch 50 | loss: 0.01644 | val_logits_ll: 0.01712 |  0:02:32s\n",
      "epoch 60 | loss: 0.01636 | val_logits_ll: 0.01689 |  0:03:01s\n",
      "epoch 70 | loss: 0.01612 | val_logits_ll: 0.01693 |  0:03:32s\n",
      "epoch 80 | loss: 0.01592 | val_logits_ll: 0.01694 |  0:04:01s\n",
      "epoch 90 | loss: 0.01563 | val_logits_ll: 0.01675 |  0:04:31s\n",
      "epoch 100| loss: 0.01532 | val_logits_ll: 0.01684 |  0:05:01s\n",
      "epoch 110| loss: 0.01526 | val_logits_ll: 0.01673 |  0:05:31s\n",
      "epoch 120| loss: 0.015   | val_logits_ll: 0.01691 |  0:06:01s\n",
      "epoch 130| loss: 0.01467 | val_logits_ll: 0.01697 |  0:06:30s\n",
      "\n",
      "Early stopping occured at epoch 137 with best_epoch = 117 and best_val_logits_ll = 0.01659\n",
      "Best weights from best epoch are automatically used!\n",
      "\u001b[33m ------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "scores_auc_all = []\n",
    "test_cv_preds = []\n",
    "\n",
    "NB_SPLITS = 10 # 7\n",
    "mskf = MultilabelStratifiedKFold(n_splits = NB_SPLITS, random_state = 0, shuffle = True)\n",
    "\n",
    "oofID_list = []\n",
    "oof_preds = []\n",
    "oof_targets = []\n",
    "scores = []\n",
    "scores_auc = []\n",
    "\n",
    "for fold_nb, (train_idx, val_idx) in enumerate(mskf.split(train_df, targets)):\n",
    "    print(b_,\"FOLDS: \", r_, fold_nb + 1)\n",
    "    print(g_, '*' * 60, c_)\n",
    "    \n",
    "    ID_train, ID_valid = train_df_ID.values[train_idx], train_df_ID.values[val_idx]\n",
    "    print(ID_valid)\n",
    "    X_train, y_train = train_df.values[train_idx, :], targets.values[train_idx, :]\n",
    "    X_val, y_val = train_df.values[val_idx, :], targets.values[val_idx, :]\n",
    "    ### Model ###\n",
    "    model = TabNetRegressor(**tabnet_params)\n",
    "        \n",
    "    ### Fit ###\n",
    "    # Another change to the original code\n",
    "    # virtual_batch_size of 32 instead of 128\n",
    "    model.fit(\n",
    "        X_train = X_train,\n",
    "        y_train = y_train,\n",
    "        eval_set = [(X_val, y_val)],\n",
    "        eval_name = [\"val\"],\n",
    "        eval_metric = [\"logits_ll\"],\n",
    "        max_epochs = MAX_EPOCH,\n",
    "        patience = 20,\n",
    "        batch_size = 1024, \n",
    "        virtual_batch_size = 32,\n",
    "        num_workers = 1,\n",
    "        drop_last = False,\n",
    "        # To use binary cross entropy because this is not a regression problem\n",
    "        loss_fn = F.binary_cross_entropy_with_logits\n",
    "    )\n",
    "    print(y_, '-' * 60)\n",
    "    \n",
    "    file = open(f'MODEL{fold_nb}.pkl', 'wb')\n",
    "    pickle.dump(model, file)\n",
    "    file.close()\n",
    "    \n",
    "    ### Predict on validation ###\n",
    "    preds_val = model.predict(X_val)\n",
    "    # Apply sigmoid to the predictions\n",
    "    preds = 1 / (1 + np.exp(-preds_val))\n",
    "    score = np.min(model.history[\"val_logits_ll\"])\n",
    "    \n",
    "    ### Save OOF for CV ###\n",
    "    oof_preds.append(preds_val)\n",
    "    oof_targets.append(y_val)\n",
    "    \n",
    "    oofID_list.append(ID_valid)\n",
    "    scores.append(score)\n",
    "    \n",
    "    ### Predict on test ###\n",
    "    preds_test = model.predict(X_test)\n",
    "    test_cv_preds.append(1 / (1 + np.exp(-preds_test)))\n",
    "\n",
    "oof_preds_all = np.concatenate(oof_preds)\n",
    "oof_targets_all = np.concatenate(oof_targets)\n",
    "test_preds_all = np.stack(test_cv_preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "oofID_all = np.concatenate(oofID_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>g-0</th>\n",
       "      <th>g-2</th>\n",
       "      <th>g-3</th>\n",
       "      <th>g-4</th>\n",
       "      <th>g-5</th>\n",
       "      <th>g-6</th>\n",
       "      <th>g-7</th>\n",
       "      <th>g-8</th>\n",
       "      <th>g-9</th>\n",
       "      <th>g-10</th>\n",
       "      <th>...</th>\n",
       "      <th>gc_mean</th>\n",
       "      <th>g_std</th>\n",
       "      <th>c_std</th>\n",
       "      <th>gc_std</th>\n",
       "      <th>g_kurt</th>\n",
       "      <th>c_kurt</th>\n",
       "      <th>gc_kurt</th>\n",
       "      <th>g_skew</th>\n",
       "      <th>c_skew</th>\n",
       "      <th>gc_skew</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.731433</td>\n",
       "      <td>-0.254104</td>\n",
       "      <td>-0.615315</td>\n",
       "      <td>-0.194236</td>\n",
       "      <td>-0.772257</td>\n",
       "      <td>-1.061067</td>\n",
       "      <td>0.005623</td>\n",
       "      <td>0.513989</td>\n",
       "      <td>-0.130963</td>\n",
       "      <td>1.094414</td>\n",
       "      <td>...</td>\n",
       "      <td>0.045182</td>\n",
       "      <td>0.600465</td>\n",
       "      <td>0.506976</td>\n",
       "      <td>0.602876</td>\n",
       "      <td>-0.006781</td>\n",
       "      <td>-0.250230</td>\n",
       "      <td>-0.025857</td>\n",
       "      <td>-0.010803</td>\n",
       "      <td>0.118667</td>\n",
       "      <td>-0.039037</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.017020</td>\n",
       "      <td>0.213820</td>\n",
       "      <td>0.027107</td>\n",
       "      <td>0.849834</td>\n",
       "      <td>0.482972</td>\n",
       "      <td>0.217779</td>\n",
       "      <td>0.363904</td>\n",
       "      <td>-0.298420</td>\n",
       "      <td>0.730029</td>\n",
       "      <td>-0.808914</td>\n",
       "      <td>...</td>\n",
       "      <td>0.045437</td>\n",
       "      <td>0.588216</td>\n",
       "      <td>0.421628</td>\n",
       "      <td>0.584621</td>\n",
       "      <td>-0.048549</td>\n",
       "      <td>0.111352</td>\n",
       "      <td>-0.096328</td>\n",
       "      <td>0.096908</td>\n",
       "      <td>-0.043717</td>\n",
       "      <td>0.003906</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.462792</td>\n",
       "      <td>0.981712</td>\n",
       "      <td>-0.091777</td>\n",
       "      <td>-0.039121</td>\n",
       "      <td>1.055526</td>\n",
       "      <td>0.161005</td>\n",
       "      <td>0.242395</td>\n",
       "      <td>0.050070</td>\n",
       "      <td>0.975415</td>\n",
       "      <td>-0.385579</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.022895</td>\n",
       "      <td>0.643043</td>\n",
       "      <td>0.458371</td>\n",
       "      <td>0.623742</td>\n",
       "      <td>-0.366360</td>\n",
       "      <td>-0.051865</td>\n",
       "      <td>-0.285094</td>\n",
       "      <td>-0.015580</td>\n",
       "      <td>0.379548</td>\n",
       "      <td>0.011925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.531916</td>\n",
       "      <td>-0.269127</td>\n",
       "      <td>0.455184</td>\n",
       "      <td>1.595325</td>\n",
       "      <td>-0.621171</td>\n",
       "      <td>-1.509230</td>\n",
       "      <td>0.206249</td>\n",
       "      <td>-0.066407</td>\n",
       "      <td>-0.889590</td>\n",
       "      <td>-0.656089</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.096822</td>\n",
       "      <td>0.722179</td>\n",
       "      <td>0.392981</td>\n",
       "      <td>0.724922</td>\n",
       "      <td>-0.877465</td>\n",
       "      <td>4.444686</td>\n",
       "      <td>-0.907331</td>\n",
       "      <td>0.108122</td>\n",
       "      <td>2.010409</td>\n",
       "      <td>0.272977</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>-0.341668</td>\n",
       "      <td>0.808567</td>\n",
       "      <td>0.604615</td>\n",
       "      <td>0.951946</td>\n",
       "      <td>-0.630950</td>\n",
       "      <td>-0.243933</td>\n",
       "      <td>-0.103294</td>\n",
       "      <td>-0.684346</td>\n",
       "      <td>0.741863</td>\n",
       "      <td>-0.175749</td>\n",
       "      <td>...</td>\n",
       "      <td>0.024536</td>\n",
       "      <td>0.765814</td>\n",
       "      <td>0.471472</td>\n",
       "      <td>0.743453</td>\n",
       "      <td>-0.353646</td>\n",
       "      <td>-0.477550</td>\n",
       "      <td>-0.243094</td>\n",
       "      <td>-0.127465</td>\n",
       "      <td>0.191181</td>\n",
       "      <td>-0.198329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 947 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        g-0       g-2       g-3       g-4       g-5       g-6       g-7  \\\n",
       "0  0.731433 -0.254104 -0.615315 -0.194236 -0.772257 -1.061067  0.005623   \n",
       "1  0.017020  0.213820  0.027107  0.849834  0.482972  0.217779  0.363904   \n",
       "2  0.462792  0.981712 -0.091777 -0.039121  1.055526  0.161005  0.242395   \n",
       "3 -0.531916 -0.269127  0.455184  1.595325 -0.621171 -1.509230  0.206249   \n",
       "4 -0.341668  0.808567  0.604615  0.951946 -0.630950 -0.243933 -0.103294   \n",
       "\n",
       "        g-8       g-9      g-10  ...   gc_mean     g_std     c_std    gc_std  \\\n",
       "0  0.513989 -0.130963  1.094414  ...  0.045182  0.600465  0.506976  0.602876   \n",
       "1 -0.298420  0.730029 -0.808914  ...  0.045437  0.588216  0.421628  0.584621   \n",
       "2  0.050070  0.975415 -0.385579  ... -0.022895  0.643043  0.458371  0.623742   \n",
       "3 -0.066407 -0.889590 -0.656089  ... -0.096822  0.722179  0.392981  0.724922   \n",
       "4 -0.684346  0.741863 -0.175749  ...  0.024536  0.765814  0.471472  0.743453   \n",
       "\n",
       "     g_kurt    c_kurt   gc_kurt    g_skew    c_skew   gc_skew  \n",
       "0 -0.006781 -0.250230 -0.025857 -0.010803  0.118667 -0.039037  \n",
       "1 -0.048549  0.111352 -0.096328  0.096908 -0.043717  0.003906  \n",
       "2 -0.366360 -0.051865 -0.285094 -0.015580  0.379548  0.011925  \n",
       "3 -0.877465  4.444686 -0.907331  0.108122  2.010409  0.272977  \n",
       "4 -0.353646 -0.477550 -0.243094 -0.127465  0.191181 -0.198329  \n",
       "\n",
       "[5 rows x 947 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "train_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CV log_loss:  0.016504302204819635\n"
     ]
    }
   ],
   "source": [
    "score = 0\n",
    "\n",
    "for i in range(len(target_cols)):\n",
    "    score += log_loss(oof_targets_all[:, i], 1 / (1 + np.exp(-oof_preds_all[:, i])))\n",
    "\n",
    "print(\"CV log_loss: \", score / y_pred.shape[1])\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "((21948, 206), (21948, 206))"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "oof_targets_all.shape,oof_preds_all.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "oof shape: (21948, 207)\n"
     ]
    }
   ],
   "source": [
    "valid_results = targets.copy()\n",
    "target_cols = targets.columns.tolist()\n",
    "for i in range(len(target_cols)):\n",
    "    valid_results[target_cols[i]] = 1 / (1 + np.exp(-oof_preds_all[:, i]))  \n",
    "    \n",
    "valid_results['sig_id'] = oofID_all\n",
    "\n",
    "print('oof shape:',valid_results.shape)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>5-alpha_reductase_inhibitor</th>\n",
       "      <th>11-beta-hsd1_inhibitor</th>\n",
       "      <th>acat_inhibitor</th>\n",
       "      <th>acetylcholine_receptor_agonist</th>\n",
       "      <th>acetylcholine_receptor_antagonist</th>\n",
       "      <th>acetylcholinesterase_inhibitor</th>\n",
       "      <th>adenosine_receptor_agonist</th>\n",
       "      <th>adenosine_receptor_antagonist</th>\n",
       "      <th>adenylyl_cyclase_activator</th>\n",
       "      <th>adrenergic_receptor_agonist</th>\n",
       "      <th>...</th>\n",
       "      <th>trpv_agonist</th>\n",
       "      <th>trpv_antagonist</th>\n",
       "      <th>tubulin_inhibitor</th>\n",
       "      <th>tyrosine_kinase_inhibitor</th>\n",
       "      <th>ubiquitin_specific_protease_inhibitor</th>\n",
       "      <th>vegfr_inhibitor</th>\n",
       "      <th>vitamin_b</th>\n",
       "      <th>vitamin_d_receptor_agonist</th>\n",
       "      <th>wnt_inhibitor</th>\n",
       "      <th>sig_id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000977</td>\n",
       "      <td>0.002040</td>\n",
       "      <td>0.001323</td>\n",
       "      <td>0.002686</td>\n",
       "      <td>0.003138</td>\n",
       "      <td>0.002175</td>\n",
       "      <td>0.003150</td>\n",
       "      <td>0.004832</td>\n",
       "      <td>0.001245</td>\n",
       "      <td>0.004631</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001717</td>\n",
       "      <td>0.003366</td>\n",
       "      <td>0.002477</td>\n",
       "      <td>0.009945</td>\n",
       "      <td>0.000854</td>\n",
       "      <td>0.017642</td>\n",
       "      <td>0.001399</td>\n",
       "      <td>0.000278</td>\n",
       "      <td>0.001920</td>\n",
       "      <td>id_002452c7e</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000175</td>\n",
       "      <td>0.000614</td>\n",
       "      <td>0.000410</td>\n",
       "      <td>0.000277</td>\n",
       "      <td>0.000412</td>\n",
       "      <td>0.000781</td>\n",
       "      <td>0.000549</td>\n",
       "      <td>0.001725</td>\n",
       "      <td>0.000124</td>\n",
       "      <td>0.000557</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001446</td>\n",
       "      <td>0.000883</td>\n",
       "      <td>0.003982</td>\n",
       "      <td>0.015393</td>\n",
       "      <td>0.000200</td>\n",
       "      <td>0.011924</td>\n",
       "      <td>0.000310</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000533</td>\n",
       "      <td>id_003603254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.001796</td>\n",
       "      <td>0.000994</td>\n",
       "      <td>0.002702</td>\n",
       "      <td>0.005273</td>\n",
       "      <td>0.003265</td>\n",
       "      <td>0.001899</td>\n",
       "      <td>0.002756</td>\n",
       "      <td>0.003805</td>\n",
       "      <td>0.000684</td>\n",
       "      <td>0.003586</td>\n",
       "      <td>...</td>\n",
       "      <td>0.002935</td>\n",
       "      <td>0.004415</td>\n",
       "      <td>0.000198</td>\n",
       "      <td>0.001102</td>\n",
       "      <td>0.001239</td>\n",
       "      <td>0.000659</td>\n",
       "      <td>0.001712</td>\n",
       "      <td>0.000126</td>\n",
       "      <td>0.001250</td>\n",
       "      <td>id_003fdd734</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000925</td>\n",
       "      <td>0.000624</td>\n",
       "      <td>0.001675</td>\n",
       "      <td>0.021982</td>\n",
       "      <td>0.037327</td>\n",
       "      <td>0.005910</td>\n",
       "      <td>0.001416</td>\n",
       "      <td>0.014271</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.003821</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000800</td>\n",
       "      <td>0.002216</td>\n",
       "      <td>0.000214</td>\n",
       "      <td>0.001166</td>\n",
       "      <td>0.000279</td>\n",
       "      <td>0.000913</td>\n",
       "      <td>0.002574</td>\n",
       "      <td>0.003327</td>\n",
       "      <td>0.001333</td>\n",
       "      <td>id_00548fd5c</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.001475</td>\n",
       "      <td>0.002224</td>\n",
       "      <td>0.001710</td>\n",
       "      <td>0.008786</td>\n",
       "      <td>0.030014</td>\n",
       "      <td>0.006707</td>\n",
       "      <td>0.002217</td>\n",
       "      <td>0.004941</td>\n",
       "      <td>0.000149</td>\n",
       "      <td>0.004756</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001414</td>\n",
       "      <td>0.003883</td>\n",
       "      <td>0.004979</td>\n",
       "      <td>0.001622</td>\n",
       "      <td>0.000871</td>\n",
       "      <td>0.001520</td>\n",
       "      <td>0.002003</td>\n",
       "      <td>0.000854</td>\n",
       "      <td>0.001288</td>\n",
       "      <td>id_006e27d96</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 207 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   5-alpha_reductase_inhibitor  11-beta-hsd1_inhibitor  acat_inhibitor  \\\n",
       "0                     0.000977                0.002040        0.001323   \n",
       "1                     0.000175                0.000614        0.000410   \n",
       "2                     0.001796                0.000994        0.002702   \n",
       "3                     0.000925                0.000624        0.001675   \n",
       "4                     0.001475                0.002224        0.001710   \n",
       "\n",
       "   acetylcholine_receptor_agonist  acetylcholine_receptor_antagonist  \\\n",
       "0                        0.002686                           0.003138   \n",
       "1                        0.000277                           0.000412   \n",
       "2                        0.005273                           0.003265   \n",
       "3                        0.021982                           0.037327   \n",
       "4                        0.008786                           0.030014   \n",
       "\n",
       "   acetylcholinesterase_inhibitor  adenosine_receptor_agonist  \\\n",
       "0                        0.002175                    0.003150   \n",
       "1                        0.000781                    0.000549   \n",
       "2                        0.001899                    0.002756   \n",
       "3                        0.005910                    0.001416   \n",
       "4                        0.006707                    0.002217   \n",
       "\n",
       "   adenosine_receptor_antagonist  adenylyl_cyclase_activator  \\\n",
       "0                       0.004832                    0.001245   \n",
       "1                       0.001725                    0.000124   \n",
       "2                       0.003805                    0.000684   \n",
       "3                       0.014271                    0.000094   \n",
       "4                       0.004941                    0.000149   \n",
       "\n",
       "   adrenergic_receptor_agonist  ...  trpv_agonist  trpv_antagonist  \\\n",
       "0                     0.004631  ...      0.001717         0.003366   \n",
       "1                     0.000557  ...      0.001446         0.000883   \n",
       "2                     0.003586  ...      0.002935         0.004415   \n",
       "3                     0.003821  ...      0.000800         0.002216   \n",
       "4                     0.004756  ...      0.001414         0.003883   \n",
       "\n",
       "   tubulin_inhibitor  tyrosine_kinase_inhibitor  \\\n",
       "0           0.002477                   0.009945   \n",
       "1           0.003982                   0.015393   \n",
       "2           0.000198                   0.001102   \n",
       "3           0.000214                   0.001166   \n",
       "4           0.004979                   0.001622   \n",
       "\n",
       "   ubiquitin_specific_protease_inhibitor  vegfr_inhibitor  vitamin_b  \\\n",
       "0                               0.000854         0.017642   0.001399   \n",
       "1                               0.000200         0.011924   0.000310   \n",
       "2                               0.001239         0.000659   0.001712   \n",
       "3                               0.000279         0.000913   0.002574   \n",
       "4                               0.000871         0.001520   0.002003   \n",
       "\n",
       "   vitamin_d_receptor_agonist  wnt_inhibitor        sig_id  \n",
       "0                    0.000278       0.001920  id_002452c7e  \n",
       "1                    0.000078       0.000533  id_003603254  \n",
       "2                    0.000126       0.001250  id_003fdd734  \n",
       "3                    0.003327       0.001333  id_00548fd5c  \n",
       "4                    0.000854       0.001288  id_006e27d96  \n",
       "\n",
       "[5 rows x 207 columns]"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "valid_results.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [],
   "source": [
    "valid_results.to_csv('oof_model2.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**The worst CV value that i achive**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = \"seagreen\">Conclusion (NOT AVAILABLE UNTIL I SEE THE LB Score)</font> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color = \"seagreen\">Submission</font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sig_id</th>\n",
       "      <th>5-alpha_reductase_inhibitor</th>\n",
       "      <th>11-beta-hsd1_inhibitor</th>\n",
       "      <th>acat_inhibitor</th>\n",
       "      <th>acetylcholine_receptor_agonist</th>\n",
       "      <th>acetylcholine_receptor_antagonist</th>\n",
       "      <th>acetylcholinesterase_inhibitor</th>\n",
       "      <th>adenosine_receptor_agonist</th>\n",
       "      <th>adenosine_receptor_antagonist</th>\n",
       "      <th>adenylyl_cyclase_activator</th>\n",
       "      <th>...</th>\n",
       "      <th>tropomyosin_receptor_kinase_inhibitor</th>\n",
       "      <th>trpv_agonist</th>\n",
       "      <th>trpv_antagonist</th>\n",
       "      <th>tubulin_inhibitor</th>\n",
       "      <th>tyrosine_kinase_inhibitor</th>\n",
       "      <th>ubiquitin_specific_protease_inhibitor</th>\n",
       "      <th>vegfr_inhibitor</th>\n",
       "      <th>vitamin_b</th>\n",
       "      <th>vitamin_d_receptor_agonist</th>\n",
       "      <th>wnt_inhibitor</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>id_0004d9e33</td>\n",
       "      <td>0.000971</td>\n",
       "      <td>0.000999</td>\n",
       "      <td>0.002072</td>\n",
       "      <td>0.020190</td>\n",
       "      <td>0.020764</td>\n",
       "      <td>0.004569</td>\n",
       "      <td>0.003068</td>\n",
       "      <td>0.007401</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000766</td>\n",
       "      <td>0.001469</td>\n",
       "      <td>0.003622</td>\n",
       "      <td>0.000977</td>\n",
       "      <td>0.000686</td>\n",
       "      <td>0.000567</td>\n",
       "      <td>0.000458</td>\n",
       "      <td>0.002147</td>\n",
       "      <td>0.002455</td>\n",
       "      <td>0.001444</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>id_001897cda</td>\n",
       "      <td>0.000595</td>\n",
       "      <td>0.000943</td>\n",
       "      <td>0.002507</td>\n",
       "      <td>0.002778</td>\n",
       "      <td>0.001552</td>\n",
       "      <td>0.001995</td>\n",
       "      <td>0.000989</td>\n",
       "      <td>0.010378</td>\n",
       "      <td>0.000821</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000821</td>\n",
       "      <td>0.001199</td>\n",
       "      <td>0.003134</td>\n",
       "      <td>0.000391</td>\n",
       "      <td>0.005183</td>\n",
       "      <td>0.000606</td>\n",
       "      <td>0.007797</td>\n",
       "      <td>0.001326</td>\n",
       "      <td>0.007307</td>\n",
       "      <td>0.002805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>id_002429b5b</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>id_00276f245</td>\n",
       "      <td>0.000987</td>\n",
       "      <td>0.000928</td>\n",
       "      <td>0.001718</td>\n",
       "      <td>0.008641</td>\n",
       "      <td>0.016314</td>\n",
       "      <td>0.003524</td>\n",
       "      <td>0.002505</td>\n",
       "      <td>0.004519</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000625</td>\n",
       "      <td>0.001723</td>\n",
       "      <td>0.002399</td>\n",
       "      <td>0.019245</td>\n",
       "      <td>0.004015</td>\n",
       "      <td>0.000570</td>\n",
       "      <td>0.001547</td>\n",
       "      <td>0.001905</td>\n",
       "      <td>0.000350</td>\n",
       "      <td>0.001453</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>id_0027f1083</td>\n",
       "      <td>0.001567</td>\n",
       "      <td>0.001331</td>\n",
       "      <td>0.001393</td>\n",
       "      <td>0.013533</td>\n",
       "      <td>0.020962</td>\n",
       "      <td>0.005249</td>\n",
       "      <td>0.004211</td>\n",
       "      <td>0.002697</td>\n",
       "      <td>0.000401</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000688</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.002572</td>\n",
       "      <td>0.002471</td>\n",
       "      <td>0.001482</td>\n",
       "      <td>0.000602</td>\n",
       "      <td>0.000720</td>\n",
       "      <td>0.001976</td>\n",
       "      <td>0.000231</td>\n",
       "      <td>0.001751</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 207 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         sig_id  5-alpha_reductase_inhibitor  11-beta-hsd1_inhibitor  \\\n",
       "0  id_0004d9e33                     0.000971                0.000999   \n",
       "1  id_001897cda                     0.000595                0.000943   \n",
       "2  id_002429b5b                     0.000000                0.000000   \n",
       "3  id_00276f245                     0.000987                0.000928   \n",
       "4  id_0027f1083                     0.001567                0.001331   \n",
       "\n",
       "   acat_inhibitor  acetylcholine_receptor_agonist  \\\n",
       "0        0.002072                        0.020190   \n",
       "1        0.002507                        0.002778   \n",
       "2        0.000000                        0.000000   \n",
       "3        0.001718                        0.008641   \n",
       "4        0.001393                        0.013533   \n",
       "\n",
       "   acetylcholine_receptor_antagonist  acetylcholinesterase_inhibitor  \\\n",
       "0                           0.020764                        0.004569   \n",
       "1                           0.001552                        0.001995   \n",
       "2                           0.000000                        0.000000   \n",
       "3                           0.016314                        0.003524   \n",
       "4                           0.020962                        0.005249   \n",
       "\n",
       "   adenosine_receptor_agonist  adenosine_receptor_antagonist  \\\n",
       "0                    0.003068                       0.007401   \n",
       "1                    0.000989                       0.010378   \n",
       "2                    0.000000                       0.000000   \n",
       "3                    0.002505                       0.004519   \n",
       "4                    0.004211                       0.002697   \n",
       "\n",
       "   adenylyl_cyclase_activator  ...  tropomyosin_receptor_kinase_inhibitor  \\\n",
       "0                    0.000221  ...                               0.000766   \n",
       "1                    0.000821  ...                               0.000821   \n",
       "2                    0.000000  ...                               0.000000   \n",
       "3                    0.000244  ...                               0.000625   \n",
       "4                    0.000401  ...                               0.000688   \n",
       "\n",
       "   trpv_agonist  trpv_antagonist  tubulin_inhibitor  \\\n",
       "0      0.001469         0.003622           0.000977   \n",
       "1      0.001199         0.003134           0.000391   \n",
       "2      0.000000         0.000000           0.000000   \n",
       "3      0.001723         0.002399           0.019245   \n",
       "4      0.000779         0.002572           0.002471   \n",
       "\n",
       "   tyrosine_kinase_inhibitor  ubiquitin_specific_protease_inhibitor  \\\n",
       "0                   0.000686                               0.000567   \n",
       "1                   0.005183                               0.000606   \n",
       "2                   0.000000                               0.000000   \n",
       "3                   0.004015                               0.000570   \n",
       "4                   0.001482                               0.000602   \n",
       "\n",
       "   vegfr_inhibitor  vitamin_b  vitamin_d_receptor_agonist  wnt_inhibitor  \n",
       "0         0.000458   0.002147                    0.002455       0.001444  \n",
       "1         0.007797   0.001326                    0.007307       0.002805  \n",
       "2         0.000000   0.000000                    0.000000       0.000000  \n",
       "3         0.001547   0.001905                    0.000350       0.001453  \n",
       "4         0.000720   0.001976                    0.000231       0.001751  \n",
       "\n",
       "[5 rows x 207 columns]"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_feat = [col for col in submission.columns if col not in [\"sig_id\"]]\n",
    "# To obtain the same lenght of test_preds_all and submission\n",
    "test = pd.read_csv(data_path + \"test_features.csv\")\n",
    "sig_id = test[test[\"cp_type\"] != \"ctl_vehicle\"].sig_id.reset_index(drop = True)\n",
    "tmp = pd.DataFrame(test_preds_all.mean(axis = 0), columns = all_feat)\n",
    "tmp[\"sig_id\"] = sig_id\n",
    "\n",
    "submission = pd.merge(test[[\"sig_id\"]], tmp, on = \"sig_id\", how = \"left\")\n",
    "submission.fillna(0, inplace = True)\n",
    "\n",
    "#submission[all_feat] = tmp.mean(axis = 0)\n",
    "\n",
    "# Set control to 0\n",
    "#submission.loc[test[\"cp_type\"] == 0, submission.columns[1:]] = 0\n",
    "submission.to_csv(\"submission_model2.csv\", index = None)\n",
    "submission.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[34msubmission.shape: \u001b[31m(3982, 207)\n"
     ]
    }
   ],
   "source": [
    "print(f\"{b_}submission.shape: {r_}{submission.shape}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class = \"alert alert-block alert-info\">\n",
    "    <h3><font color = \"red\">NOTE: </font></h3>\n",
    "    <p>If you want to comment please tag me with '@' to answer more quickly.</p>\n",
    "</div>"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "31f2aee4e71d21fbe5cf8b01ff0e069b9275f58929596ceb00d14d90e3e16cd6"
  },
  "kernelspec": {
   "display_name": "Python 3.8.9 64-bit",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
